File: Chapter 07/JS/section_7_4_event_code/shared/orderPlaced.js
import { Event } from './event.js';
import crypto from 'crypto';

/**
 * THE POCO: An immutable fact representing a completed action.
 * In Event-Driven Architecture, events are named in the past tense because 
 * they represent things that have already happened. 
 */
export class OrderPlaced extends Event {
    /**
     * Constructs the specific data payload for when an order is placed.
     * @param {string} correlationId - The thread tying this back to the user's web request.
     * @param {string} userId - The user who placed the order.
     * @param {number} totalAmount - The total cost.
     */
    constructor(correlationId, userId, totalAmount) {
        super(correlationId);
        this.orderId = crypto.randomUUID();
        this.userId = userId;
        this.totalAmount = totalAmount;
        
        // We use Object.freeze in vanilla JS to simulate the strict immutability 
        // found in C# records. An event is a historical fact; it must never be altered 
        // by downstream consumers.
        Object.freeze(this); 
    }
}

File: Chapter 07/Python/section_7_4_event_code/shared/order_placed.py
# shared/order_placed.py

from dataclasses import dataclass, field
from datetime import datetime
import uuid
from .event import Event

@dataclass(frozen=True) 
class OrderPlaced(Event):
    """
    THE POCO: An immutable fact representing a completed action.
    
    In Event-Driven Architecture, events are named in the past tense because 
    they represent things that have already happened. 
    
    We use Python's @dataclass(frozen=True) decorator to guarantee immutability. 
    An event is a historical fact; once created, it must never be altered by 
    downstream consumers. If it could be altered, we would lose our system's 
    single source of truth.
    """
    correlation_id: uuid.UUID
    order_id: uuid.UUID
    user_id: uuid.UUID
    total_amount: float
    
    # We use default_factory so these are generated dynamically upon instantiation,
    # rather than being shared across all instances of the class.
    event_id: uuid.UUID = field(default_factory=uuid.uuid4)
    occurred_on: datetime = field(default_factory=datetime.utcnow)